File: graph/bipartite.py
# import numpy as np
from collections import deque

class bipartite_graph:

    def __init__(self,n,m):
        self.n = n
        self.m = m
        self.g = [list() for i in range(n+m)]
    
    def add_edge(self,v,u):
        if u < v: v,u = u,v
        assert v >= 0 and v < self.n and u >= self.n and u < self.n + self.m
        self.g[v].append(u)
        self.g[u].append(v)

    def _augment(self, v, match):
        if(match[v] >= 0):
            return False
        
        par = [-2]*(n+m) # np.full(n+m, -2, int)
        q = deque()
        par[v] = -1
        q.append(v)
        found = False

        while(len(q) > 0):
            v = q.pop()
            for u in [z for z in self.g[v] if par[z] == -2]:
                par[u] = v
                if(match[u] < 0):
                    found = True
                    break
                elif(par[match[u]] == -2):
                    par[match[u]] = u
                    q.append(match[u])
            if(found):
                break
 
        if(not found):
            return False
        
        # augment match
        while(True):
            match[u] = v
            tmp = match[v]
            match[v] = u
            if(tmp >= 0):
                u = tmp
                v = par[u]
            else:
                break
        return True

    def max_matching(self):
        ans = 0
        match = [-1]*(n+m) # np.full(n+m, -1, int)
        for i in range(self.n):
            if(self._augment(i, match)):
                ans += 1
        return ans


if __name__ == "__main__":
    # https://www.hackerrank.com/challenges/real-estate-broker/problem
    
    n,m = map(int, input().split())

    client = [list(map(int, input().split())) for _ in range(n)]
    house = [list(map(int,input().split())) for _ in range(m)]

    g = bipartite_graph(n,m)
    for i in range(n):
        for j in range(m):
            if(client[i][0] < house[j][0] and client[i][1] >= house[j][1]):
                g.add_edge(i, j+n)

    print(g.max_matching())

File: graph/bipartite.cpp
#include <bits/stdc++.h>

using namespace std;
#define mp make_pair
#define all(x) (x).begin(), (x).end()
typedef vector<int> vi;
typedef pair<int, int> pii;
typedef long long ll;

class bipartite_graph
{
    /* 
        Bipartide graph with n vertices on the left and m vertices on the right.
        Index with [0,n-1] on the left and [n,n+m-1] on the right.
    */
   
public:

    vector<vector<int>> g;
    int n, m;

    bipartite_graph(int n, int m) : n(n), m(m)
    {
        for (int i = 0; i < n + m; i++)
            g.push_back(vi());
    }

    void add_edge(int v, int u)
    {
        if(u < v) swap(u,v);
        assert(v >= 0 && v < n && u >= n && u < n+m);

        g[v].push_back(u);
        g[u].push_back(v);
    }

    bool augment(int v, vector<bool> &used, vector<int> &match)
    {
        used[v] = true;
        for (int u : g[v])
        {
            if (used[u]) continue;
            
            used[u] = true;
            bool found_path = false;
            if (match[u] < 0)
            {
                found_path = true;
            }else if(!used[match[u]]){
                int vv = match[u];
                if(augment(vv, used, match)){
                    found_path = true;
                }
            }
            if(found_path){
                match[u] = v;
                match[v] = u;
                return true;
            }
        }
        return false;
    }

    int max_matching()
    {
        vector<bool> used(n + m);
        vector<int> match(n+m, -1);

        int ans = 0;
        for (int i = 0; i < n; i++)
        {
            fill(all(used), false);
            if (augment(i, used, match))
                ans++;
            if(ans == min(n,m))
                break;
        }
        return ans;
    }

    string to_string(){
        stringstream s;
        s << "n = " << n << " m = " << m << endl;
        for(int i=0;i<n;i++){
            s << "N(" << i << ") = ";
            for(auto u : g[i])
                s << u << " ";
            s << endl;
        }
        return s.str();
    }
};

int main()
{
    // https://www.hackerrank.com/challenges/real-estate-broker/problem
    int n, m;

    cin >> n >> m;

    vector<pii> client(n);
    vector<pii> house(m);

    int i,j,x,y;

    for(i=0;i<n;i++){
        cin >> x >> y;
        client[i] = mp(x,y);
    }
    for(i=0;i<m;i++){
        cin >> x >> y;
        house[i] = mp(x,y);
    }
    
    
    bipartite_graph g(n,m);

    for(i=0;i<n;i++){
        for(j=0;j<m;j++){
            if(client[i].first < house[j].first && client[i].second >= house[j].second){
                g.add_edge(i, j+n);
            }
        }
    }

    //cout << g.to_string() << endl;

    cout << g.max_matching() << endl;

    return 0;
}